### PR TITLE
WIP: Bounds checking (PR 2)

### PR DESCRIPTION
**Pull request by ryco117**
##### Description
Applied checking that the Ethereum-network matches between the browser and bounty for fulfilling, increasing, killing, and processing bounties. Added check to ensure an owner of a bounty cannot fulfill one in their possession. Added deadline and stage check for fulfilling bounty as well as stage check for processing bounty. Also factored out a shared function `string web3NetworkIdToString(EthNetId)` for easy retrieving of a pretty name for current given network id, a shared enum for stages, and a shared function browserNetworkIs(networkStr, callbackFn(bool){}).

##### Checklist
- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)

js/pages

##### Testing
Manually created, fulfilled, increasing, and processed bounties on my local copy (in addition to testing that the new boundaries were enforced.)

##### Refers/Fixes

Fixes: #445 

This is a PR since I kinda botched the commit history of the previous fork..

### CODE REVIEW

**Comment by owocki** [comment]
network id checks seem to work on mainnet / rinkeby, but dont work on localhost, kovan, or ropsten

File: app/assets/v2/js/pages/fulfill_bounty.js [hunk 1]
```diff
@@ -149,15 +149,53 @@ window.onload = function() {
...
+                // creator cannot fulfill their own bounty.
+                var fromAddress = result['bounty_owner_address'];
+
+                if (fromAddress == account) {
+                  _alert({ message: 'The address that funded an issue cannot fulfill it.' });
+                  unloading_button($('.js-submit'));
+                  return;
+                }
+
+                browserNetworkIs(bountyNetwork, function(matchingNetworks) {
```
Comment (owocki) [inline app/assets/v2/js/pages/fulfill_bounty.js]: hey @mbeacom @thelostone-mc @tra38 @KamesCG  you all are good with javascript... is there a way we can do multiple checks against the blockchain without all these async callbacks?  the nested callbacks make it really hard for me to follow whats going on here.

File: app/assets/v2/js/pages/kill_bounty.js [hunk 2]
```diff
@@ -60,60 +60,68 @@ window.onload = function() {
...
+                  document.location.href = '/funding/details?url=' + issueURL;
+                }, 1000);
+              };
+
+              if (error) {
+                mixpanel.track('Kill Bounty Error', {
+                  step: 'final_callback',
+                  error: error
+                });
+                console.error('err', error);
```
Comment (owocki) [inline app/assets/v2/js/pages/kill_bounty.js]: errant console statement!

File: app/assets/v2/js/shared.js [hunk 3]
```diff
@@ -647,6 +647,31 @@ var trigger_form_hooks = function() {
   trigger_faucet_form_web3_hooks();
 };
 
+var web3NetworkIdToString = function(netId) {
```
Comment (owocki) [inline app/assets/v2/js/shared.js]: thx for the refacotr here!

File: app/assets/v2/js/pages/process_bounty.js [hunk 4]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+            var next = function() {
+              // setup inter page state
+              localStorage[issueURL] = JSON.stringify({
+                'timestamp': timestamp(),
+                'dataHash': null,
+                'issuer': account,
+                'txid': result
+              });
+
+              _alert({ message: 'Submitted transaction to web3.' }, 'info');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Any reason for the omission of `gettext` from some of the alert messages?

**Comment by ryco117** [comment]
@owocki, yes; some of the error messages used that function and some did not before my commit. Because it was not clear that gettext provided any specific function (when the result is already a string) I didn't make an effort to wrap my errors with it.

**Comment by owocki** [comment]
@mbeacom ^^

File: app/assets/v2/js/pages/increase_bounty.js [hunk 5]
```diff
@@ -105,53 +105,78 @@ $(document).ready(function() {
...
+        var errormsg = undefined;
+
+        if (bountyAmount == 0 || open == false) {
+          errormsg = gettext('No active funded issue found at this address. Are you sure this is an active funded issue?');
+        }
+        if (fromAddress != web3.eth.coinbase) {
+          errormsg = gettext('Only the address that submitted this funded issue may increase the payout.');
+        }
+        if (!matchingNetworks) {
+          errormsg = 'Expected browser to be connected to the Ethereum network' +
```
Comment (mbeacom) [inline app/assets/v2/js/pages/increase_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/increase_bounty.js [hunk 6]
```diff
@@ -105,53 +105,78 @@ $(document).ready(function() {
...
-    }
+    // Use DB to determine which networks each bounty is on.
+    var uri = '/api/v0.1/bounties/?github_url=' + issueURL;
+
+    $.get(uri, function(results, status) {
+      results = sanitizeAPIResults(results);
+      result = results[0];
+      if (result == null) {
+        _alert({
+          message: 'No active bounty found for this Github URL.'
```
Comment (mbeacom) [inline app/assets/v2/js/pages/increase_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 7]
```diff
@@ -78,11 +78,11 @@ window.onload = function() {
         localStorage['acceptTOS'] = true;
       }
       if (issueURL == '') {
-        _alert({ message: gettext('Please enter a issue URL.') }, 'warning');
+        _alert({ message: 'Please enter a issue URL.' }, 'warning');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please re-wrap this in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 8]
```diff
@@ -78,11 +78,11 @@ window.onload = function() {
...
         isError = true;
       }
       if (fulfillmentId == null) {
-        _alert({ message: gettext('Please enter a fulfillment Id.') }, 'warning');
+        _alert({ message: 'Please enter a fulfillment Id.' }, 'warning');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please re-wrap this in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 9]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
       var apiCallback = function(results, status) {
         if (status != 'success') {
           mixpanel.track('Process Bounty Error', {step: 'apiCallback', error: error});
-          _alert({ message: gettext('Could not get bounty details') }, 'warning');
+          _alert({ message: 'Could not get bounty details' }, 'warning');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please re-wrap this in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 10]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
           console.error(error);
           unloading_button($('.submitBounty'));
           return;
         }
         results = sanitizeAPIResults(results);
         result = results[0];
         if (result == null) {
-          _alert({ message: gettext('No active bounty found for this Github URL.') }, 'info');
+          _alert({ message: 'No active bounty found for this Github URL.' }, 'info');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please re-wrap this in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 11]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+          var fromAddress = result['bounty_owner_address'];
+          var claimeeAddress = result['fulfiller_address'];
+          var open = result['is_open'];
+          var initialized = true;
+          var bountyId = result['standard_bounties_id'];
+
+          var errormsg = undefined;
+
+          if (bountyAmount == 0 || open == false || initialized == false) {
+            errormsg = 'No active funding found at this address.  Are you sure this is an active funded issue?';
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 12]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+          } else if (claimeeAddress == '0x0000000000000000000000000000000000000000') {
+            errormsg = 'No claimee found for this bounty.';
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 13]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+          } else if (fromAddress != web3.eth.coinbase) {
+            errormsg = 'You can only process a funded issue if you submitted it initially.';
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 14]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+          } else if (!matchingNetworks) {
+            errormsg = 'Expected browser to be connected to the Ethereum network' +
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js (hunk 4 above)
Comment (mbeacom) [reply app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 15]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+              setTimeout(function() {
+                mixpanel.track('Process Bounty Success', {});
+                document.location.href = '/funding/details?url=' + issueURL;
+              }, 1000);
+
+            };
+
+            if (error) {
+              mixpanel.track('Process Bounty Error', {step: 'final_callback', error: error});
+              _alert({ message: 'There was an error' }, 'error');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

File: app/assets/v2/js/pages/process_bounty.js [hunk 16]
```diff
@@ -96,72 +96,87 @@ window.onload = function() {
...
+              unloading_button($('.submitBounty'));
+            } else {
+              next();
+            }
+          };
 
+          // Retrieve parameters for bounty from blockchain
+          bounty.getBounty.call(bountyId, (errStr, bountyParams) => {
+            if (bountyParams[4] != bountyStageEnum['Active']) {
+              _alert({ message: 'The bounty for this Github URL is not active.' });
```
Comment (mbeacom) [inline app/assets/v2/js/pages/process_bounty.js]: Please wrap this string in `gettext()`

**Comment by owocki** [comment]
this seems good to merge once @mbeacom 's comments are addressed and once the merge conflicts are resolved

**Comment by owocki** [comment]
> @owocki, yes; some of the error messages used that function and some did not before my commit. Because it was not clear that gettext provided any specific function (when the result is already a string) I didn't make an effort to wrap my errors with it.

ahh okay.. yeah its a django translation method

**Comment by ryco117** [comment]
Will make the suggested changes, thanks for the patience guys!!

**Comment by ryco117** [comment]
@owocki About your comment on it not working for other networks, I will admit to it only working on those in my testing, however, the gitcoin UI itself was restricting me to just those two choices before my changes (correct me if I'm wrong on this, or if theres been a change)

**Comment by owocki** [comment]
hmmm somehow theres a bunch of commits from myself and mbeacon on this PR now

File: app/assets/v2/js/pages/fulfill_bounty.js (hunk 1 above)
Comment (thelostone-mc) [reply app/assets/v2/js/pages/fulfill_bounty.js]: We could do Promises / async-await but we'd need ES6 for that  😓

**Comment by mbeacom** [comment]
Closing this for now since it's been a while.
Please feel free to re-open if you'd like to resolve the merge conflicts and rebase issues @ryco117 - Looking forward to hearing from you!

Thanks again

**Pull request closed without merging**